File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.js
function solution(edges) {    
    const graphInfo = {}

    edges.forEach(([send,receive])=>{
        graphInfo[send] ||= { sendCount:0, receiveCount:0 }
        graphInfo[receive] ||= { sendCount:0, receiveCount:0 }

        graphInfo[send].sendCount++
        graphInfo[receive].receiveCount++
    })

    let s = 0
    let count = 0
    let bar = 0, donut = 0, eight = 0

    for (let v in graphInfo){
        const { sendCount, receiveCount } = graphInfo[v]
        if (sendCount >= 2 && receiveCount === 0){
            s = +v
            count = sendCount
        }
        if (sendCount == 0) bar++
        if (sendCount >= 2 && receiveCount >= 2) eight++
    }

    donut = count - eight - bar

    return [s, donut, bar, eight];
}

File: 프로그래머스/2/258711. 도넛과 막대 그래프/도넛과 막대 그래프.py
from collections import defaultdict, deque

def solution(edges):
    answer = [0, 0, 0, 0]
    start, donut, bar, eight = 0, 1, 2, 3
    graph = defaultdict(list)
    degrees = defaultdict(lambda: [0, 0])

    for edge in edges:
        a, b = edge
        graph[a].append(b)

    for node, nodes in graph.items():
        for adjacent in nodes:
            degrees[node][0] += 1
            degrees[adjacent][1] += 1

    for node, degree in degrees.items():
        outd, ind = degree

        if outd == 0: answer[bar] += 1
        elif outd == 2:
            if ind > 0: answer[eight] += 1
            else:
                answer[start] = node
                root = node
        elif outd > 2:
            answer[start] = node
            root = node

    answer[donut] = degrees[root][0] - answer[bar] - answer[eight]

    return answer